### PR TITLE
Add 3.0 changelog

### PR DESCRIPTION
**Pull request by raddevon**
This is mostly done, but I have some outstanding questions.

1. I'm not sure about most of my instructions for upgrading hosted instances. I got some guidance from @elprans (how to add repo for testing under Debian/Ubuntu, what the installation commands need to be), but I just made up the changes needed for CentOS/RHEL and the `systemctl` commands. I think I have the Docker tag correct, although that one isn't live yet so I can't confirm.
2. Seems like, based on the feedback I'm getting re: 3.0 readiness across clients, I could probably just omit the "Client libraries" section for now? The two I've heard back on so far seem to be very close but not live yet.
3. Does anyone see anything in the "Bug fixes" section that doesn't belong? Anything not there that should be? There were 600+ PRs to go through starting from the first 3.0 PR. I did my best to eliminate anything that was backported or related to a feature and to leave only the items that sounded like clear bug fixes, but I undoubtedly didn't get it right 100% of the time.

### CODE REVIEW

File: docs/changelog/3_x.rst [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+This process will involve some downtime, specifically during steps 2 and 3.
+
+
+Client libraries
+----------------
+
+We've released new versions of our JavaScript and Python client libraries that
+support all 3.0 features and implement the updated protocol. These versions
+are backwards compatible with v1.x and v2.x instances, so we encourage all users to
+upgrade.
```
Comment (1st1) [inline docs/changelog/3_x.rst:101]: Let's mention that many client libraries gained codegen capabilities, since the release of 2.0

File: docs/changelog/3_x.rst [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+      trigger log_insert after insert for each do (
+        insert Log {
+          action := 'insert',
+          target_name := __new__.name
+        }
+      );
+    }
+
+The trigger above inserts a ``Log`` object any time a ``Person`` object is
+inserted.
```
Comment (1st1) [inline docs/changelog/3_x.rst:181]: Add a link to https://github.com/edgedb/rfcs/blob/master/text/1020-triggers.rst

File: docs/changelog/3_x.rst [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+      required body: str;
+      modified: datetime {
+        rewrite insert, update using (datetime_of_statement())
+      }
+    }
+
+This shows one reason mutation rewrites is one of our most wanted features:
+modified timestamps! When a user inserts or updates a ``Post``, the rewrite
+will set the value of the ``modified`` property to that value of
+``datetime_of_statement()``. There are tons of other uses too. Give them a try!
```
Comment (1st1) [inline docs/changelog/3_x.rst:205]: Add a link to https://github.com/edgedb/rfcs/blob/master/text/1021-rewrites.rst

File: docs/changelog/3_x.rst [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+          },
+          default::Villain {
+            id: 335f4104-b777-11ed-81eb-ab4de34e9c36,
+            name: 'Obadiah Stane'
+          },
+        },
+      },
+    }
+
+It's a super-handy way to quickly explore your data.
```
Comment (1st1) [inline docs/changelog/3_x.rst:277]: Add a link to https://github.com/edgedb/rfcs/blob/master/text/1023-splats.rst

File: docs/changelog/3_x.rst [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+in any way that makes sense to you.
+
+.. code-block:: sdl
+
+    module MommaModule {
+      module BabyModule {
+        <schema-declarations>
+      }
+    }
+
```
Comment (1st1) [inline docs/changelog/3_x.rst:297]: Mention that this will allow us to expand the list of standard modules in the future in a backwards compatible way.

Comment (1st1) [reply docs/changelog/3_x.rst:297]: Also need to add an example of how this looks like in EdgeQL

File: docs/changelog/3_x.rst [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+Among other improvements, the UI now includes a visual query planner to help
+you tweak performance on your EdgeQL queries. Just drop the ``analyze`` keyword
+in front of your query in the UI's "Query Editor" tab to see the query planner
+in action.
+
+.. image:: images/v3_ui.jpg
+    :width: 100%
+
+Query planning is available in the CLI REPL, also by prepending your query with
+``analyze``.
```
Comment (1st1) [inline docs/changelog/3_x.rst:331]: This needs to be elevated to right under the syntax change

File: docs/changelog/3_x.rst [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+
+* Implement schema reset (:eql:gh:`#4714`)
+
+* Support link properties on computed backlinks (:eql:gh`#5227`)
+
+
+CLI
+---
+
+* Add the ``edgedb migration edit`` command (:ref:`docs
```
Comment (1st1) [inline docs/changelog/3_x.rst:389]: What about devmode?

File: docs/changelog/3_x.rst [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+  <ref_cli_edgedb_migration_edit>`)
+
+
+Server
+------
+
+* Improvements to top-level server error reporting (:eql:gh:`#5349`)
+
+  * For MultiErrors reported from task groups, don't lose
+    chained exceptions
```
Comment (1st1) [inline docs/changelog/3_x.rst:399]: I don't think this should be mentioned

File: docs/changelog/3_x.rst [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+  * When an InternalServerError kills the server, produce
```
Comment (1st1) [inline docs/changelog/3_x.rst:400]: IMO this has been implemented since 1.0

File: docs/changelog/3_x.rst [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+    the "please file a bug" message
+
+
+
+Bug fixes
```
Comment (1st1) [inline docs/changelog/3_x.rst:405]: We also need to add a section about UI, which was massively redesigned:

* New way of setting globals / configs
* New "Terminal"
* The Editor panel was completely redesigned, with history & visual query builder
* Mention the new explain there too

File: docs/changelog/3_x.rst [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+    - ``edgedb@0.4.0``
+
+New features
+============
+
+Simplified SDL syntax
+---------------------
+
+As part of our commitment to delivering the best developer experience in
+databases, we've made our schema definition language (or SDL) easier to use.
```
[the remark this answers is not in the document]
Comment (Dhghomon) [reply docs/changelog/3_x.rst:126]: > I kinda like this, but on the other hand, the other very important reason was us making this change to enable future syntax for declaring ad-hoc types in queries and functions. I'd mention that. Please also link this: https://github.com/edgedb/rfcs/blob/master/text/1022-freetypes.rst

Ah, so that's why! I had been wondering if there was a benefit besides a slicker syntax.

File: docs/changelog/3_x.rst [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+    }
+
+As you're reading through our docs, you'll notice that changing the version
+selector in the sidebar will also switch all SDL examples to the appropriate
+SDL syntax.
+
+.. note::
+
+    SDL syntax in documentation does not change for non-versioned content like
+    guides.
```
Comment (Dhghomon) [inline docs/changelog/3_x.rst:156]: I'm reading this as "SDL syntax in documentation for non-versioned content like guides will take longer to upgrade to the new syntax", is that right? i.e. probably updated on a best-effort basis but won't be 3.0 syntax starting today.

If that's so I might reword to:

"SDL in documentation like guides is non-versioned content and will be updated to the new syntax over time."

File: docs/changelog/3_x.rst [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+will set the value of the ``modified`` property to that value of
+``datetime_of_statement()``. There are tons of other uses too. Give them a try!
+
+
+Splats
+------
+
+This is one of the most fun features in 3.0, both to say _and_ to use! With
+splats, you can easily select all properties in your queries without typing all
+of them out.
```
Comment (Dhghomon) [inline docs/changelog/3_x.rst:213]: Would be nice to show just how awesome the splat operator is with a before/after comparison so new readers are fully aware. Something like this:

What you had to type before: 
select Movie { id, release_year, title };

Now:
select Movie {*};

What you had to type before:
select Movie { 
  id, 
  release_year,
  title,
  characters: {
    id,
    name
  }
}

Now:
select Movie {**};

File: docs/changelog/3_x.rst (hunk 7 above)
Comment (tailhook) [reply docs/changelog/3_x.rst:389]: Also `migration edit` was it 2.1. It's probably fine to mention (as there is no blog post for CLI 2.1), but we have to mark it as such I think.

File: docs/changelog/3_x.rst [hunk 14]
```diff
@@ -0,0 +1,517 @@
...
+      select <optional tuple<name: str, flag: bool>>$var;
+
+* Implement migration rewrites (:eql:gh:`#4585`)
+
+* Implement schema reset (:eql:gh:`#4714`)
+
+* Support link properties on computed backlinks (:eql:gh`#5227`)
+
+
+CLI
```
Comment (tailhook) [inline docs/changelog/3_x.rst:386]: * `edgedb watch`
* `edgedb migrate --dev-mode`
* `edgedb migration upgrade-check`
  * at some point we should not forget that `upgrade` now executes `upgrade-check` (but not yet)
* `edgedb migration create --squash`
* `edgedb analyze`, `\analyze`, `--expand`
* `edgedb info --get` 2.1 maybe worth mentioning
`--channel=testing, --to-testing` in `instance create`, `project init` and `cli upgrade`
* `\d object xxx` -> `\d xxx` maybe worth mentioning?
* Cloud support?

File: docs/changelog/3_x.rst [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+        echo deb [signed-by=/usr/local/share/keyrings/edgedb-keyring.gpg] \
+          https://packages.edgedb.com/apt \
+          $(grep "VERSION_CODENAME=" /etc/os-release | cut -d= -f2) testing \
+          | sudo tee /etc/apt/sources.list.d/edgedb.list
+
+    Use this command for installation under Debian/Ubuntu:
+
+    .. code-block:: bash
+
+        sudo apt-get update && sudo apt-get install edgedb-3
```
[the remark this answers is not in the document]
Comment (raddevon) [reply docs/changelog/3_x.rst:60]: This was just a guess on my part. I have no idea what it should be. You're most likely correct.

Quick question: why `beta2` instead of `beta1`?

[the remark this answers is not in the document]
Comment (tailhook) [reply docs/changelog/3_x.rst]: Sure `beta1`.

File: docs/changelog/3_x.rst (hunk 1 above)
Comment (raddevon) [reply docs/changelog/3_x.rst]: I'd like to link to the documentation for code generation in the libraries that gained it too. I see documentation on code generation for TS, Python, and Dart. Did TS already have code generation as of 2.0, or is it also new for this version? Any others I need to mention besides those?

File: docs/changelog/3_x.rst (hunk 12 above)
Comment (raddevon) [reply docs/changelog/3_x.rst:144]: That's sorta what I'm trying to say here. I really just want people to know that changing the version dropdown won't affect SDL code in non-versioned content.

Maybe I can write around this by changing the preceding paragraph. Would do you think about this rewrite of that paragraph, @Dhghomon?

> Selecting "v3" from the version dropdown in the sidebar will update any SDL code in versioned sections of the documentation to the new syntax.

Some of the guides and other non-versioned content may be "polished up" for v3. Others may need more substantial rewrites. Some may be retired instead if they aren't widely used. Some may live on as v2 guides in perpetuity. We haven't really ironed out the details around exactly what that transition is going to look like, so I'm wary of making any promises with respect to that content.

File: docs/changelog/3_x.rst [hunk 16]
```diff
@@ -0,0 +1,517 @@
...
+.. code-block:: sdl
+
+    module MommaModule {
+      module BabyModule {
+        <schema-declarations>
+      }
+    }
+
+
+``intersect`` and ``except`` operators
```
[the remark this answers is not in the document]
Comment (raddevon) [reply docs/changelog/3_x.rst:329]: As far as I know, these do not. I think these are set operators. @vpetrovykh, can you confirm?

Just tested, and they do _something_ with free objects, but I'm having trouble reasoning about what.

```
edgedb> select {name := 'Devon', age := 40} intersect {name := 'Devon', age := 41};
{}
```

```
edgedb> select {name := 'Devon', age := 40} except {name := 'Devon', age := 41};
{{id: a4d77281-bdb2-40dc-b93a-de4eebb85a76}}
```

File: docs/changelog/3_x.rst (hunk 12 above)
Comment (Dhghomon) [reply docs/changelog/3_x.rst:144]: Yeah, that looks good and then no need to talk about guides and such. Then maybe just remove the "any" so it doesn't sound like 100% coverage? (Might be missed in a few corners somewhere..)

File: docs/changelog/3_x.rst [hunk 17]
```diff
@@ -372,14 +372,56 @@ Use ``except`` to find members of the first set that are not in the second.
...
+          assert(
+            not exists (__new__.friends intersect __new__.enemies),
+            message := "Invalid frenemies",
+          )
+        )
+      }
+
+.. code-block:: edgeql-repl
+
+    db> insert Person {name := 'Quincey Morris'};
```
Comment (Dhghomon) [inline docs/changelog/3_x.rst:371]: Nice example! 🧛 and `intersect` at the same time

File: docs/changelog/3_x.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (tailhook) [reply docs/changelog/3_x.rst:329]: Yes, free objects are weird. But all types are sets. So something like this works:
```
edgedb> with a := (select schema::Object), b := (select schema::Object FILTER .name like 's%')
....... select (a except b) {name}
```
Not sure what is a good example, though.

File: docs/changelog/3_x.rst (hunk 14 above)
Comment (raddevon) [reply docs/changelog/3_x.rst:527]: Thank you for this list, @tailhook. I just added `watch` and the development workflow — an abridged version of what I wrote previously for the documentation. I also added mention of the `analyze` commands to the section on query planning, although I don't know what `--expand` is or how to use it. I'll circle back now and pick up the testing channel stuff and the new syntax of the describe object backslash command.

Can you fill me in on these?

* `edgedb migrate --dev-mode`
* `--expand option`

My thinking on Cloud support is that, even though we have already added it, it's not relevant to mention until Cloud launches. Open to hear arguments against that approach.

Comment (raddevon) [reply docs/changelog/3_x.rst:543]: Just to spell out what I need to know:

* `edgedb migrate --dev-mode` isn't documented on `--help` best I can tell. I created a new migration and ran it and there was no output. Can't really tell what it does.
* I'm not sure what command `--expand` is an option for.

File: docs/changelog/3_x.rst [hunk 18]
```diff
@@ -197,6 +197,57 @@ even when you're just learning EdgeQL.
...
+
+
+3. Edit your schema files
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As your application evolves, directly edit your schema files to reflect your
+desired data model. When you save your changes, ``watch`` will immediately
+begin applying your new schema to the database.
+
+Once you have the schema the way you want it, and you're ready to lock it in
```
Comment (tailhook) [inline docs/changelog/3_x.rst:228]: "lock it in by generating migration"
Or
"lock it in (see below)"

Otherwise it's hard to figure out what you mean and the start of the next section looks like entirely unrelated thing, until you read the whole section.

**Review by tailhook — APPROVED** [review]
Other than nit below, looks good.

**Comment by raddevon** [comment]
@tailhook Did you want something about `edgedb migrate --dev-mode` and `--expand` in here? I had some questions about them if so.

File: docs/changelog/3_x.rst [hunk 19]
```diff
@@ -0,0 +1,729 @@
...
+   In any required ``systemctl`` commands, replace ``edgedb-server-2`` with
+   ``edgedb-server-3``.
+
+   Under any Docker setups, supply the ``3.0-beta.1`` tag.
+
+2. Take your application offline, then dump your v2.x database with the CLI
+
+   .. code-block:: bash
+
+       $ edgedb dump --dsn <old dsn> --all my_database.dump/
```
Comment (zackelan) [inline docs/changelog/3_x.rst:77]: do we need `--format=dir` here? this is in the help output:

```
OPTIONS:
        --all                Dump all databases and the server configuration. path is a
                             directory in this case
        --format <format>    Choose dump format. For normal dumps this parameter should be
                             omitted. For --all only --format=dir is required.
                              [possible values: dir]
```

File: docs/changelog/3_x.rst [hunk 20]
```diff
@@ -0,0 +1,729 @@
...
+
+   This will dump the schema and contents of your current database to a file
```
Comment (zackelan) [inline docs/changelog/3_x.rst:79]: directory, not file

File: docs/changelog/3_x.rst [hunk 21]
```diff
@@ -0,0 +1,729 @@
...
+  .. code-block:: edgeql
+
+      select <tuple<str, bool>>$var;
+      select <optional tuple<str, bool>>$var;
+      select <tuple<name: str, flag: bool>>$var;
+      select <optional tuple<name: str, flag: bool>>$var;
+
+* Add the syntax for abstract indexes (:eql:gh:`#4691`)
+
+  Exposes some Postgres indexes that you can use in your schemas. These are
```
Comment (zackelan) [inline docs/changelog/3_x.rst:514]: could link to the Postgres docs here for people who want more details about the fancy index types:

https://www.postgresql.org/docs/current/indexes-types.html

**Comment by vpetrovykh** [comment]
Looks good.

File: docs/changelog/3_x.rst (hunk 19 above)
Comment (raddevon) [reply docs/changelog/3_x.rst]: Good catch!

**Comment by msullivan** [comment]
Should we mention that upgrades from the beta to RCs and final reversions will be "smooth" and not require a dump/restore? Might encourage people to try it more

File: docs/changelog/3_x.rst [hunk 22]
```diff
@@ -0,0 +1,732 @@
...
+  $ edgedb migration create
+
+This "locks in" the changes you prototyped using the ``watch`` command. Now,
+these are ready to commit and push to your remote to share with your team.
+
+
+Triggers
+--------
+
+Triggers is one of the most anticipated 3.0 features! They allow you to define
```
Comment (msullivan) [inline docs/changelog/3_x.rst:255]: triggers are

File: docs/changelog/3_x.rst [hunk 23]
```diff
@@ -0,0 +1,732 @@
...
+      {Json("{\"hello\": \"world\"}")}
+
+* Support tuples as query arguments (:eql:gh:`#4489`)
+
+  .. code-block:: edgeql
+
+      select <tuple<str, bool>>$var;
+      select <optional tuple<str, bool>>$var;
+      select <tuple<name: str, flag: bool>>$var;
+      select <optional tuple<name: str, flag: bool>>$var;
```
Comment (msullivan) [inline docs/changelog/3_x.rst:510]: Maybe an example with an array?

File: docs/changelog/3_x.rst [hunk 24]
```diff
@@ -15,14 +15,57 @@ To play with the new features, install the CLI using
 
   $ edgedb project init --server-version=testing
 
+.. note::
+
+    Good news, everyone! Upgrades from the betas to release candidates (RCs)
+    and to the final release will *not* require a dump and restore. 🎉 You can try
+    out the beta with the assurance that this upgrade to the beta will be the
+    last dump and restore that is required of you.
```
Comment (msullivan) [inline docs/changelog/3_x.rst:23]: ... for 3.0 :P

Comment (raddevon) [reply docs/changelog/3_x.rst]: Circled back to make it extra clear we're only talking about 3.0 here.

File: docs/changelog/3_x.rst [hunk 25]
```diff
@@ -0,0 +1,726 @@
...
+
+   Once the restore is complete, update your application to connect to the new
+   instance.
+
+   This process will involve some downtime, specifically during steps 2 and 3.
+
+**Pre-1.0 Instances**
+
+If you're still running pre-1.0 EdgeDB instances (e.g., 1.0-beta3) and want
+to upgrade to 3.0, you would need to upgrade to version 2.x first, followed by
```
Comment (elprans) [inline docs/changelog/3_x.rst:140]: I'd say we *recommend* you upgrade to 2.x, but it shouldn't be necessary, unless I missed some specific incompatibility.

File: docs/changelog/3_x.rst [hunk 26]
```diff
@@ -0,0 +1,726 @@
...
+.. code-block:: sdl
+
+    type User {
+      required email: str;
+      multi friends: User;
+    }
+
+Selecting "v3" from the version dropdown in the sidebar will update SDL code in
+versioned sections of the documentation to the new syntax.
+
```
Comment (elprans) [inline docs/changelog/3_x.rst:188]: I would mention that the old syntax is still fully supported.

File: docs/changelog/3_x.rst [hunk 27]
```diff
@@ -0,0 +1,726 @@
...
+
+Query plan
+----------
+
+Among other improvements, the UI now includes a visual query planner to help
```
Comment (elprans) [inline docs/changelog/3_x.rst:193]:
```suggestion
Among other improvements, the UI now includes a visual query analyzer to help
```

File: docs/changelog/3_x.rst [hunk 28]
```diff
@@ -0,0 +1,726 @@
...
+you tweak performance on your EdgeQL queries. Just drop the ``analyze`` keyword
+in front of your query in the UI's "Query Editor" tab to see the query planner
+in action.
+
+.. image:: images/v3_ui_query_planner.jpg
+    :width: 100%
+
+Query planning is available in the CLI REPL by prepending your query with
```
Comment (elprans) [inline docs/changelog/3_x.rst:201]:
```suggestion
Query analysis is available in the CLI REPL by prepending your query with
```

File: docs/changelog/3_x.rst [hunk 29]
```diff
@@ -0,0 +1,726 @@
...
+Selecting "v3" from the version dropdown in the sidebar will update SDL code in
+versioned sections of the documentation to the new syntax.
+
+
+Query plan
+----------
+
+Among other improvements, the UI now includes a visual query planner to help
+you tweak performance on your EdgeQL queries. Just drop the ``analyze`` keyword
+in front of your query in the UI's "Query Editor" tab to see the query planner
```
Comment (elprans) [inline docs/changelog/3_x.rst:195]:
```suggestion
in front of your query in the UI's "Query Editor" tab to see the query analyzer
```

File: docs/changelog/3_x.rst (hunk 25 above)
Comment (raddevon) [reply docs/changelog/3_x.rst:140]: Hey, @fantix. Does "recommend" work here, or is this upgrade path a requirement?

File: docs/changelog/3_x.rst (hunk 27 above)
Comment (raddevon) [reply docs/changelog/3_x.rst:193]: @elprans Should the title of this section also change from "Query plan" to "Query analysis" or "Query analyzer?"

Comment (elprans) [reply docs/changelog/3_x.rst:193]: Query performance analysis

File: docs/changelog/3_x.rst [hunk 30]
```diff
@@ -0,0 +1,726 @@
+====
```
Comment (1st1) [inline docs/changelog/3_x.rst:1]: Can you mark it as a "new in 3.0" kind of page?

Comment (1st1) [reply docs/changelog/3_x.rst:1]: Because it's not yet released

File: docs/changelog/3_x.rst [hunk 31]
```diff
@@ -0,0 +1,726 @@
...
+v3.0
+====
+
+:edb-alt-title: EdgeDB v3
```
Comment (1st1) [inline docs/changelog/3_x.rst:5]: Add a warning saying that the latest stable is 2.x and 3.0 is expected to be released soon

File: docs/changelog/3_x.rst (hunk 30 above)
Comment (raddevon) [reply docs/changelog/3_x.rst:3]: Is this what you intended, @1st1?
image

I've done it, but it's a little weird since it now responds to the version selector even though it's outside the versioned fence.

File: docs/changelog/3_x.rst (hunk 25 above)
Comment (fantix) [reply docs/changelog/3_x.rst:147]: Yes sure, we only dropped the test coverage to restore pre-1.0 dumps. So it should still be available now to restore directly into 3.0 without having to go through 2.x, until some day in the future we introduce such incompatibility.

File: docs/changelog/3_x.rst (hunk 30 above)
Comment (1st1) [reply docs/changelog/3_x.rst:3]: Got it. Let's just change the title of the page to `EdgeDB v3 (dev)` and the TOC entry to `v3.0 (dev)`? (keeping the filename as is)

File: docs/changelog/3_x.rst [hunk 32]
```diff
@@ -0,0 +1,738 @@
...
+Read more about splats in `our splats RFC
+<https://github.com/edgedb/rfcs/blob/master/text/1023-splats.rst>`_.
+
+
+Nested modules
+--------------
+
+"Yo, dawg. We heard you like modules…" (Sorry. We've been working so hard on
+3.0, `we got way way behind on our memes
+<https://knowyourmeme.com/memes/xzibit-yo-dawg>`_.)
```
Comment (1st1) [inline docs/changelog/3_x.rst:417]: I'm not sure I'm a huge fan of this particular joke :) Maybe drop it?

**Comment by 1st1** [comment]
Thank you, Devon! I've merged it and will trigger a deploy now.
